lib/sycamore/sycamore/data/table.py

```diff
@@ -0,0 +1,187 @@
+from dataclasses import dataclass, field
+from typing import Any, Optional, TypeVar, Union
+import xml.etree.ElementTree as ET
+
+from sycamore.data import BoundingBox
+from PIL import Image, ImageDraw
+import numpy as np
+from pandas import DataFrame
+
+
+# This is part of itertools in 3.10+.
+# Adding here to support 3.9
+def _pairwise(iterable):
+    iterator = iter(iterable)
+    a = next(iterator, None)
+    for b in iterator:
+        yield a, b
+        a = b
+
+
+# This data model is similar to that used by Textract and TableTransformers.
+# It is intended to be a common intermediate representation for a variety of
+# table structure recognition models.
+@dataclass(frozen=True)
+class TableCell:
+    content: str
+    rows: list[int]
+    cols: list[int]
+    is_header: bool = False
+    bbox: Optional[BoundingBox] = None
+    # Model/format specific properties
+    properties: dict[str, Any] = field(default_factory=dict)
+
+    def __post_init__(self):
+        # Validate that row and column spans are contiguous.
+        for a, b in _pairwise(self.rows):
+            if a + 1 != b:
+                raise ValueError(f"Found non-contiguous rows in {self}.")
+
+        for a, b in _pairwise(self.cols):
+            if a + 1 != b:
+                raise ValueError(f"Found non-contiguous cols in {self}.")
+
+
+class Table:
+    def __init__(self, cells: list[TableCell], caption: Optional[str] = None):
+        self.cells: list[TableCell] = sorted(cells, key=lambda tc: (min(tc.rows), min(tc.cols)))
+        self.caption = caption
+        self.num_rows = max(max(c.rows) for c in self.cells) + 1
+        self.num_cols = max(max(c.cols) for c in self.cells) + 1
+
+    def __eq__(self, other):
+        if type(other) is not type(self):
+            return False
+
+        if self.num_rows != other.num_rows or self.num_cols != other.num_cols:
+            return False
+
+        # TODO: Yikes! this is expensive. Do the sorting in init or something.
+        if self.cells != other.cells:
+            return False
+
+        return True
+
+    def __hash__(self):
+        return hash((self.cells))
+
+    # This algorithm is modified from the TableTransformers code. The conversion to Pandas/CSV
+    # is necessarily lossy since these formats requires tables to be square and don't support
+    # cells spanning multiple rows/columns.
+    #
+    # One of the main decisions here is how to render cells that span multiple rows and columns,
+    # since they are not nativey supported in CSV. The standard approaches are (1) to duplicate
+    # the cell value in each row/column or (2) to put the value in the first row/column in the cell
+    # and empty values in the other. TableTransformers do the former, The textractor library supports
+    # both, but defaults to the latter. Since our primary use case is preparing context for LLMs,
+    # we speculate that duplication may create confusion, so we default to only displaying a cells
```

Comment:
How the LLM impact our decision here? You mean duplicates cause problem? This matters with the format we send to LLM in either delimited csv format or html format, seems the later one has more info loss.
Besides, do you feel it make sense to provide both modes by taking some knob?

Reply:
I thought I convinced myself that duplicates might be more confusing – for example imagine you wanted to sum up the cells in a column, but you duplicated values – you would get the wrong answer. Not completely sure though. I was just trying to pick something, and this is at least consistent with what we get from textract today. 

I think you lose less structural information when you convert to html vs csv, but LLMs may be less able to deal with complex html. Not sure. I'd like to do a proper benchmark, eventually. 

Happy to add an option. My instinct is that very few people are going to want to tweak configuration parameters at this level, but if we see significantly different behavior it might be worth it. Are you okay if we defer the knob to a later commit?

Reply:
sure.